[MLIR][CF] Avoid collapsing blocks which participate in cycles

benwu25 opened the pull request:
This PR addresses #159743.

Previously, collapseBranch did not return failure for successor blocks
which were part of a cycle. mlir-opt --canonicalize would run indefinitely
for any N-block cycle which is kicked off with an unconditional jump. The
simplifyPassThroughBr transform would continue alternating which block was
targeted in ^bb0, resulting in an infinite loop.

collapseBranch will not result in any useful transformation on blocks which
participate in cycles, since the block is aliased by a different block. To
avoid this, we can check for cycles in collapseBranch and abort when one is
detected. Simplification of the cycle is left for other transforms.

Could someone help review and merge? Thanks!

benwu25 committed:
[MLIR][CF] Avoid collapsing blocks which participate in cycles (#159743)

Previously, collapseBranch did not return failure for successor blocks
which were part of a cycle. mlir-opt --canonicalize would run indefinitely
for any N-block cycle which is kicked off with an unconditional jump. The
simplifyPassThroughBr transform would continue alternating which block was
targeted in ^bb0, resulting in an infinite loop.

collapseBranch will not result in any useful transformation on blocks which
participate in cycles, since the block is aliased by a different block. To
avoid this, we can check for cycles in collapseBranch and abort when one is
detected. Simplification of the cycle is left for other transforms.

benwu25 committed:
[MLIR][CF] Update control-flow-cycles test (#159743)

benwu25 committed:
Merge branch 'main' into main

benwu25 committed:
[MLIR][CF] Fix canonicalizer regression with 1-block cycle (#159743)

I caused a regression in flang due to my previous commits, since I
did not consider encountering a single-block cycle. I updated my
check to look for 1-block cycles in addition to any larger cycles
containing the successor block.

benwu25 committed:
Merge branch 'main' into main

benwu25 commented:
@matthias-springer @zero9178

matthias-springer commented on mlir/lib/Dialect/ControlFlow/IR/ControlFlowOps.cpp:
```diff
@@ -122,6 +122,18 @@ static LogicalResult collapseBranch(Block *&successor,
+  // Don't try to collapse branches which participate in a cycle.
+  Block *currBlock = successorDest;
+  BranchOp nextBranch = dyn_cast<BranchOp>(currBlock->getTerminator());
+  while (nextBranch) {
```
Is it possible that this `while` loop runs into an endless loop? E.g.:

```
   A  ->  B  -> C  <-
                |     \
                v      \
                D   ->  E
```

Assuming that we call `collapseBranch(A, ...)`.

I feel like `nextBranchDest == successor` is not sufficient and we need a `DenseSet<Block *> visited` instead?

benwu25 replied:
I see your argument, but weirdly enough, I have already written this testcase (`@delayed_3_cycle`), and somehow it terminates. I think that `A -> B -> C` is first canonicalized as just `C` by the `simplifyBrToBlockWithSinglePred`, so then when we collapse on `C` it somehow works out.

However, if `B` has another predecessor, maybe the loop will be broken.

matthias-springer commented on mlir/test/Transforms/control-flow-cycles.mlir:
```diff
@@ -0,0 +1,93 @@
+// RUN: mlir-opt --canonicalize %s | FileCheck %s
+
+// Test that control-flow cycles are not simplified infinitely.
```
All these test cases should be in `test/Dialect/ControlFlow/canonicalize.mlir`.

benwu25 committed:
[MLIR][CF] Use a DenseSet to detect more cycles (#159743)

We need to use a DenseSet in case we encounter cycles further in the
chain of control flow which have not been simplified by other cf
transforms.

benwu25 committed:
Merge branch 'main' into main

matthias-springer commented on mlir/lib/Dialect/ControlFlow/IR/ControlFlowOps.cpp:
```diff
@@ -122,6 +122,18 @@ static LogicalResult collapseBranch(Block *&successor,
+    if (nextBranchDest == successor)
+      return failure();
+    else if (visited.contains(nextBranchDest))
+      break;
```
Also, is the above `if (nextBranchDest == successor)` still necessary if you initialize the `visited` set with `successor`?

benwu25 replied:
I chose `break` to allow for more merges. The only truly bad case seems to be when `successor` itself is part of the cycle. I.e., it is only absolutely necessary to bail out when this is the case. Consider the testcase from above:
```
func.func @cycle_4_blocks(%c : i1) {
  cf.cond_br %c, ^bb6, ^bb7
  ^bb6:
  cf.br ^bb5 {F}
  ^bb5:
  cf.br ^bb1 {A}
  ^bb1:
    cf.br ^bb2 {B}
  ^bb2:
    cf.br ^bb3 {C}
  ^bb3:
    cf.br ^bb4 {D}
  ^bb4:
    cf.br ^bb1 {E}
  ^bb7:
    cf.br ^bb6
}
```
With the prior implementation, this testcase gets simplified to:
```
func.func @cycle_4_blocks(%c : i1) {
  cf.br ^bb1
  ^bb1:
  cf.br ^bb1 {E}
}
```
But if we `return failure()`, the result is:
```
func.func @cycle_4_blocks(%c : i1) {
  cf.cond_br %c, ^bb1, ^bb3
  ^bb1:  // 2 preds: ^bb0, ^bb3
  cf.br ^bb2 {A}
  ^bb2:  // 2 preds: ^bb1, ^bb2
  cf.br ^bb2 {E}
  ^bb3:  // pred: ^bb0
  cf.br ^bb1
}
```
I believe this is because a few canonicalize passes succeed in merging blocks if we choose `break` before the cycle itself is reached.

However, it doesn't matter much if infinite loops are fully optimized, and it is simpler and still correct to just bail out, so I will add that change.

benwu25 replied:
Update:

After adding this change to bail out early, I caused a regression in flang: `test/Lower/OpenMP/infinite-loop-in-construct.f90`.

I broke this test with my initial PR because there was an infinite loop. Now, it seems like the infinite loop is less optimized than expected for this testcase. I suppose I could fix this by updating this testcase to expect a different infinite loop?

benwu25 committed:
[MLIR][CF] Bail out of collapseBranch when any cycle is detected (#159743)

It is safest to bail out as soon as we detect a cycle in collapseBranch.
We need not continue with optimizations on infinite loops.

benwu25 committed:
Merge branch 'main' into main

benwu25 committed:
[MLIR][CF] Do not bail out early in collapseBranch to avoid regression (#159743)

Bailing out as soon as a cycle is detected causes a regression in flang, where
the test expects more optimizations than we allowed on an infinite loop. We
may be able to adjust this testcase.

benwu25 committed:
Merge branch 'main' into main

benwu25 commented on mlir/test/Dialect/ControlFlow/canonicalize.mlir:
```diff
@@ -490,3 +490,139 @@ func.func @branchCondProp(%arg0: i1) {
+// CHECK:           cf.br ^bb1
+func.func @unsimplified_cycle_1(%c : i1) {
+  cf.cond_br %c, ^bb1, ^bb2
+  ^bb1:
```
Done.

benwu25 committed:
[MLIR][CF] Update broken test (#159743)

I forgot to update this test with my previous commit.
(e8135aabd5fe3a909819af29c655bd65fc3c1873)

benwu25 committed:
Merge branch 'main' into main

benwu25 committed:
[MLIR][CF] Abort collapseBranch for any cycle (#159743)

We can bail out of collapseBranch upon detecting any cycle, but we
must update a test in flang to prevent a regression.

benwu25 committed:
Merge branch 'main' into main
